solr/core/src/java/org/apache/solr/core/CoreContainer.java

```diff
@@ -944,6 +946,14 @@ public void load() {
       });
 
       clusterSingletons.setReady();
+      if (NodeRoles.PREFERRED.equals(nodeRoles.getRoleMode(NodeRoles.Role.OVERSEER))) {
+        try {
+          log.info("This node is started as a preferred overseer");
```

Comment:
Do we want role-specific if-then-else code in CoreContainer? I think there were some questions in the SIP discussion regarding where role logic should reside, and this does not seem very pluggable nor scalable.

What if NodeRole was an interface and each role was an implementation. Then CoreContainer could call some init method on each configured role when the node starts up, and that class would then do whatever it wants wrt reading its config, logging, talking to zk etc?

Reply:
Given the goal of a role is to influence behavior (here related to Overseer, elsewhere to placing or not placing data, in the future other aspects), I don't see how to make this pluggable. Sure the definition of the role can be pluggable, but to make the role be able to actually change the behavior of Solr, existing classes have to be patched.

What you suggest @janhoy is possible (discovering/initializing all roles and letting them do that work), but the missing bit is then for roles that want to change some specific behavior aspects of SolrCloud or interact with it in a different way than tweaking its configuration.

Reply:
@janhoy 

I agree with @murblanc . It's hard to have a generic interface to perform the logic of a role. The place in which this code is executed is very important. it may also alter the behavior of existing classes. However, it may be possible to have a generic interface for roles for certain roles that will come up in the future. The current design doesn't stop us from doing so

Reply:
If we are at a point where we want pluggable roles, we can easily refactor the code to achieve that. I think it is still early enough to know all usecases that roles can be used for in order to arrive at the right abstractions. All of this is internal code as of now, so no backcompat will break when we introduce a refactoring.